discovery/sync_manager.go

```diff
@@ -0,0 +1,673 @@
+	useActive bool) *GossipSyncer {
+
+	eligible := func(s *GossipSyncer) bool {
+		// Skip any syncers that exist within the blacklist.
+		if blacklist != nil {
+			if _, ok := blacklist[s.cfg.peerPub]; ok {
+				return false
+			}
+		}
+
+		// Only syncers in a chansSynced state are viable for sync
+		// transitions, so skip any that aren't.
+		return s.syncState() == chansSynced
+	}
+
+	for _, s := range m.inactiveSyncers {
+		if !eligible(s) {
+			continue
+		}
+		return s
+	}
+
+	if useActive {
+		for _, s := range m.activeSyncers {
+			if !eligible(s) {
+				continue
+			}
+			return s
+		}
+	}
+
+	return nil
+}
+
+// InitSyncState is called by outside sub-systems when a connection is
+// established to a new peer that understands how to perform channel range
+// queries. We'll allocate a new gossip syncer for it, and start any goroutines
+// needed to handle new queries.
+//
+// TODO(wilmer): Only mark as ActiveSync if this isn't a channel peer.
+func (m *SyncManager) InitSyncState(peer lnpeer.Peer) {
+	// If we already have a syncer, then we'll exit early as we don't want
+	// to override it.
+	nodeID := routing.Vertex(peer.PubKey())
+	if _, ok := m.GossipSyncer(nodeID); ok {
+		return
+	}
+
+	log.Infof("Creating new GossipSyncer for peer=%x", nodeID[:])
+
+	encoding := lnwire.EncodingSortedPlain
+	s := newGossipSyncer(gossipSyncerCfg{
+		chainHash:     m.cfg.ChainHash,
+		peerPub:       nodeID,
+		channelSeries: m.cfg.ChanSeries,
+		encodingType:  encoding,
+		chunkSize:     encodingTypeToChunkSize[encoding],
+		sendToPeer: func(msgs ...lnwire.Message) error {
+			return peer.SendMessage(false, msgs...)
+		},
+	})
+
+	// Gossip syncers are initialized by default as passive and in a
+	// chansSynced state so that they can reply to any peer queries or
+	// handle any sync transitions.
+	s.setSyncType(PassiveSync)
+	s.setSyncState(chansSynced)
+	s.Start()
+
+	m.Lock()
+	m.inactiveSyncers[nodeID] = s
+	numActiveSyncers := len(m.activeSyncers) + len(m.pendingActiveSyncers)
+	m.Unlock()
+
+	switch {
+	// If we don't have any active syncers, we'll force a historical sync
+	// with the first peer we connect to, to ensure we get as much of the
+	// graph as possible. This is mostly done for light clients, which
+	// shouldn't be receiving new updates anyway due to the high bandwidth
+	// consumption that stems from it.
+	case m.cfg.NumActiveSyncers == 0:
+		var err error
+		m.historicalSync.Do(func() {
+			log.Infof("Attempting historical sync with "+
+				"GossipSyncer(%x)", s.cfg.peerPub)
+			err = s.historicalSync()
+		})
+		if err != nil {
+			log.Errorf("Unable to perform historical sync with "+
+				"GossipSyncer(%x): %v", s.cfg.peerPub, err)
+
+			// Reset historicalSync to ensure it is tried again with
+			// a different peer.
+			m.historicalSync = sync.Once{}
+		}
+
+	// If we've yet to reach our desired number of active syncers, then
+	// we'll use this one.
+	case numActiveSyncers < m.cfg.NumActiveSyncers:
+		m.queueActiveSyncer(s)
```

Comment:
nice, this is much easier to reason about!

I think we can make it even clearer: just queue the syncer here, and let the `syncerHandler` decide whether we should do a historical sync with the syncer.

Reply:
I thought of doing this as well but seemed like there would be little to no gain since we'd still require acquiring the locks.

Reply:
Gotcha, the same pattern remains but the concurrency issue you pointed out should be fixed now.